rust/agama-lib/src/storage/client.rs

```diff
@@ -332,17 +209,35 @@ impl<'a> StorageClient<'a> {
         Ok(result)
     }
 
-    async fn build_device_info(
-        &self,
-        object: &(
-            OwnedObjectPath,
-            HashMap<OwnedInterfaceName, HashMap<std::string::String, OwnedValue>>,
-        ),
-    ) -> Result<DeviceInfo, ServiceError> {
+    fn get_interface<'b>(
+        &'b self,
+        object: &'b DBusObject,
+        name: &str,
+    ) -> Option<&HashMap<String, OwnedValue>> {
+        let interface: OwnedInterfaceName = InterfaceName::from_str_unchecked(name).into();
         let interfaces = &object.1;
-        let interface: OwnedInterfaceName =
-            InterfaceName::from_static_str_unchecked("org.opensuse.Agama.Storage1.Device").into();
-        let properties = interfaces.get(&interface);
+        interfaces.get(&interface)
+    }
+
+    async fn build_device(&self, object: &DBusObject) -> Result<Device, ServiceError> {
+        Ok(Device {
+            block_device: self.build_block_device(object).await?,
```

Comment:
The conversion of other objects (e.g., `ProposalSettings`) is done in the model by implementing the `TryFrom` trait. For the devices, the objects are directly created in the client. Probably it would be better to follow the same approach and implement `TryFrom` for the needed types in the model.